### DISCUSSION TITLE
Understanding the .notation file/model

### DISCUSSION THREAD

**Discussion by srcnuzn** [eclipse-glsp/glsp#1065] [General] [answered]
Hi.

I'm working on a custom EMF based source model by using the [java-emf-eclipse](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-eclipse) template as reference. 

I have successfully replaced the task-list model with my own EMF model.

Currently, I am working on visualizing ports for each node.

To achieve this, I looked into the .notation files and the corresponding [model classes](https://github.com/eclipse-glsp/glsp-server/tree/master/plugins/org.eclipse.glsp.server.emf/src-gen/org/eclipse/glsp/server/emf/model/notation) to get a better understanding.

1. Is it correct, that my custom source model + the provided notation model are both used to create the gmodel which is then passed to the client for visualization?
2. As far as I can see, the notation model only persists node and edge information. However, how do I add, for example, port positions to the notation model (e.g., if I want the port positions to be adaptable by the user)? The same question goes for label positions.
3. Is the notation model meant only as a reference that is to be replaced by a custom notation model, or should we actually use and extend it? 

Best regards,
srcnuzn

**Comment by planger** [comment, marked as answer]
Hi @srcnuzn,

please find my answers below.

> 1. Is it correct, that my custom source model + the provided notation model are both used to create the gmodel which is then passed to the client for visualization?

Yes, that's correct.

> 2. As far as I can see, the notation model only persists node and edge information. However, how do I add, for example, port positions to the notation model (e.g., if I want the port positions to be adaptable by the user)? The same question goes for label positions.

You can either reuse `Shape` for that or you extend the notation model. The given notation model is the simplest possible notation model that suffices for the basic use cases. However, how you use it, extend it, and how you transform it alongside a semantic model into the GModel is entirely up to you -- as implemented in your custom `GModelFactory`.

> 3. Is the notation model meant only as a reference that is to be replaced by a custom notation model, or should we actually use and extend it?

That's entirely up to you. It is used in a number of projects, often it is also extended with additional classes. If it fits your needs, feel free to use or extend it. If you prefer another way of persisting your graphical information, you can also certainly create your own.
Another alternative is to extend your semantic model with custom properties so that it can capture the graphical information too. This way things get simpler, but you "pollute" your semantic model with graphical information, which may or may not be a problem depending on your use case.

Hope that helps!
